File: packages/core/src/codegen.ts
#!/usr/bin/env node
import { execFile } from "node:child_process";
import { promisify } from "node:util";
import { writeFileSync, mkdirSync, readdirSync } from "node:fs";
import { join, resolve } from "node:path";

const exec = promisify(execFile);

const JS_RESERVED = new Set([
  "abstract", "arguments", "await", "boolean", "break", "byte", "case", "catch",
  "char", "class", "const", "continue", "debugger", "default", "delete", "do",
  "double", "else", "enum", "eval", "export", "extends", "false", "final",
  "finally", "float", "for", "function", "goto", "if", "implements", "import",
  "in", "instanceof", "int", "interface", "let", "long", "native", "new", "null",
  "package", "private", "protected", "public", "return", "short", "static",
  "super", "switch", "synchronized", "this", "throw", "throws", "transient",
  "true", "try", "typeof", "var", "void", "volatile", "while", "with", "yield",
]);

const ANSIBLE_TO_TS: Record<string, string> = {
  str: "string",
  string: "string",
  bool: "boolean",
  int: "number",
  float: "number",
  path: "string",
  raw: "unknown",
  jsonarg: "unknown",
  json: "unknown",
  dict: "Record<string, unknown>",
  list: "unknown[]",
  bits: "number",
  bytes: "number",
  sid: "string",
};

function safeIdentifier(name: string): string {
  return JS_RESERVED.has(name) ? `${name}_` : name;
}

function toPascal(name: string): string {
  return name.split("_").map(w => w.charAt(0).toUpperCase() + w.slice(1)).join("");
}

function tsType(option: Record<string, unknown>): string {
  const choices = option.choices as unknown[] | undefined;
  if (choices && choices.every(c => typeof c === "string")) {
    return choices.map(c => `"${c}"`).join(" | ");
  }

  const ansibleType = (option.type as string) ?? "str";

  if (ansibleType === "list") {
    const elements = option.elements;
    let inner: string;
    if (typeof elements === "string") {
      inner = ANSIBLE_TO_TS[elements] ?? "unknown";
    } else if (typeof elements === "object" && elements !== null) {
      inner = ANSIBLE_TO_TS[(elements as Record<string, string>).type ?? "str"] ?? "unknown";
    } else {
      inner = "unknown";
    }
    return `${inner} | ${inner}[]`;
  }

  if (ansibleType === "dict" && option.suboptions) {
    return formatSuboptions(option.suboptions as Record<string, Record<string, unknown>>);
  }

  return ANSIBLE_TO_TS[ansibleType] ?? "unknown";
}

function formatSuboptions(suboptions: Record<string, Record<string, unknown>>, indent = 2): string {
  const pad = " ".repeat(indent);
  const lines = ["{"];
  for (const [name, opt] of Object.entries(suboptions).sort(([a], [b]) => a.localeCompare(b))) {
    const suffix = opt.required ? "" : "?";
    lines.push(`${pad}  ${name}${suffix}: ${tsType(opt)};`);
  }
  lines.push(`${pad}}`);
  return lines.join("\n");
}

function generateInterface(name: string, options: Record<string, Record<string, unknown>> | undefined): string {
  if (!options || Object.keys(options).length === 0) {
    return `export type ${name} = Record<string, unknown>;`;
  }

  const lines = [`export interface ${name} {`];
  for (const [optName, opt] of Object.entries(options).sort(([a], [b]) => a.localeCompare(b))) {
    const suffix = opt.required ? "" : "?";
    const t = tsType(opt);
    const rawDesc = Array.isArray(opt.description) ? (opt.description[0] ?? "") : (opt.description ?? "");
    const desc = (rawDesc as string).replace(/\*\//g, "* /");
    if (desc) lines.push(`  /** ${desc} */`);
    const key = /^[a-zA-Z_$][a-zA-Z0-9_$]*$/.test(optName) ? optName : `"${optName}"`;
    lines.push(`  ${key}${suffix}: ${t};`);
  }
  lines.push("}");
  return lines.join("\n");
}

async function runAnsibleDoc(fqcn: string): Promise<Record<string, unknown>> {
  const { stdout } = await exec("ansible-doc", ["-t", "module", fqcn, "--json"]);
  return JSON.parse(stdout);
}

async function listCollectionModules(namespace: string): Promise<string[]> {
  const { stdout } = await exec("ansible-doc", ["-t", "module", "-l", "--json"], { maxBuffer: 50 * 1024 * 1024 });
  const all = JSON.parse(stdout) as Record<string, unknown>;
  return Object.keys(all).filter(k => k.startsWith(namespace + ".")).sort();
}

function generateModule(fqcn: string, docData: Record<string, unknown>, importPath: string): string {
  const doc = (docData[fqcn] ?? docData[Object.keys(docData)[0]!]) as Record<string, unknown>;
  const docInfo = (doc.doc ?? {}) as Record<string, unknown>;
  const returnInfo = (doc.return ?? {}) as Record<string, Record<string, unknown>> | undefined;

  const shortName = fqcn.split(".").pop()!;
  const pascal = toPascal(shortName);
  const fnName = safeIdentifier(shortName);

  const options = docInfo.options as Record<string, Record<string, unknown>> | undefined;
  const argsInterface = generateInterface(`${pascal}Args`, options);
  const returnInterface = generateInterface(`${pascal}Return`, returnInfo);

  return `// Auto-generated from: ${fqcn}
// DO NOT EDIT — regenerate with codegen

import { defineModule } from "${importPath}";

${argsInterface}

${returnInterface}

export const ${fnName} = defineModule<${pascal}Args, ${pascal}Return>("${fqcn}");
`;
}

function generateBarrel(dir: string): void {
  const files = readdirSync(dir)
    .filter(f => f.endsWith(".ts") && f !== "index.ts")
    .map(f => f.replace(/\.ts$/, ""))
    .sort();

  const lines = [
    "// Auto-generated barrel — DO NOT EDIT",
    "",
    ...files.map(f => `export * from "./${f}.js";`),
    "",
  ];
  writeFileSync(join(dir, "index.ts"), lines.join("\n"));
}

async function main() {
  const args = process.argv.slice(2);

  if (args.length === 0) {
    console.error("Usage: codegen <fqcn|namespace> [module...]");
    console.error("  codegen ansible.builtin.debug       — generate one module");
    console.error("  codegen ansible.builtin              — generate all in namespace");
    console.error("  codegen community.general debug copy — generate specific modules");
    process.exit(1);
  }

  const target = args[0]!;
  const parts = target.split(".");

  // Single fully-qualified module name (3+ parts like ansible.builtin.debug)
  if (parts.length >= 3) {
    const namespace = parts.slice(0, 2).join(".");
    const outputDir = resolveOutputDir(namespace);
    mkdirSync(outputDir, { recursive: true });

    console.error(`  Generating ${target}...`);
    const docData = await runAnsibleDoc(target);
    const importPath = resolveImportPath(namespace);
    const source = generateModule(target, docData, importPath);
    const shortName = parts.pop()!;
    writeFileSync(join(outputDir, `${shortName}.ts`), source);
    generateBarrel(outputDir);
    console.error(`Generated 1 module in ${outputDir}`);
    return;
  }

  // Namespace (2 parts like ansible.builtin)
  const namespace = target;
  const explicitModules = args.slice(1);
  const outputDir = resolveOutputDir(namespace);
  mkdirSync(outputDir, { recursive: true });

  let modules: string[];
  if (explicitModules.length > 0) {
    modules = explicitModules.map(m => m.includes(".") ? m : `${namespace}.${m}`);
  } else {
    console.error(`Listing modules for ${namespace}...`);
    modules = await listCollectionModules(namespace);
  }

  if (modules.length === 0) {
    console.error(`No modules found for namespace: ${namespace}`);
    process.exit(1);
  }

  const importPath = resolveImportPath(namespace);
  let count = 0;
  for (const fqcn of modules) {
    const shortName = fqcn.split(".").pop()!;
    console.error(`  Generating ${fqcn}...`);
    try {
      const docData = await runAnsibleDoc(fqcn);
      const source = generateModule(fqcn, docData, importPath);
      writeFileSync(join(outputDir, `${shortName}.ts`), source);
      count++;
    } catch (e) {
      console.error(`  WARNING: skipping ${fqcn}: ${e}`);
    }
  }

  generateBarrel(outputDir);
  console.error(`Generated ${count} modules in ${outputDir}`);
}

function resolveOutputDir(namespace: string): string {
  const coreDir = resolve(import.meta.dirname!, "..");
  if (namespace.startsWith("ansible.")) {
    return join(coreDir, "builtins");
  }
  return join(coreDir, "..", "community", "modules");
}

function resolveImportPath(namespace: string): string {
  if (namespace.startsWith("ansible.")) return "@sensible-ts/core";
  return "@sensible-ts/core";
}

main();

File: packages/codegen/codegen.py
#!/usr/bin/env python3
"""Generate TypeScript module bindings from ansible-doc JSON output."""

import json
import keyword
import subprocess
import sys
from pathlib import Path
from type_map import ts_type

REPO_ROOT = Path(__file__).parent.parent.parent
GENERATED_DIR = REPO_ROOT / "generated"

JS_RESERVED = {
    "abstract", "arguments", "await", "boolean", "break", "byte", "case", "catch",
    "char", "class", "const", "continue", "debugger", "default", "delete", "do",
    "double", "else", "enum", "eval", "export", "extends", "false", "final",
    "finally", "float", "for", "function", "goto", "if", "implements", "import",
    "in", "instanceof", "int", "interface", "let", "long", "native", "new", "null",
    "package", "private", "protected", "public", "return", "short", "static",
    "super", "switch", "synchronized", "this", "throw", "throws", "transient",
    "true", "try", "typeof", "var", "void", "volatile", "while", "with", "yield",
}


def safe_identifier(name: str) -> str:
    if name in JS_RESERVED or keyword.iskeyword(name):
        return f"{name}_"
    return name


def run_ansible_doc(fqcn: str) -> dict:
    result = subprocess.run(
        ["ansible-doc", "-t", "module", fqcn, "--json"],
        capture_output=True,
        text=True,
        check=True,
    )
    return json.loads(result.stdout)


def list_collection_modules(namespace: str) -> list[str]:
    result = subprocess.run(
        ["ansible-doc", "-t", "module", "-l", "--json"],
        capture_output=True,
        text=True,
        check=True,
    )
    all_modules = json.loads(result.stdout)
    return sorted(k for k in all_modules if k.startswith(namespace + "."))


def to_pascal(name: str) -> str:
    return "".join(w.capitalize() for w in name.split("_"))


def namespace_to_dir(namespace: str) -> Path:
    """ansible.builtin -> packages/core/modules, community.* -> packages/community/modules"""
    parts = namespace.split(".")
    if parts[0] == "ansible":
        return REPO_ROOT / "packages" / "core" / "builtins"
    return REPO_ROOT / "packages" / "community" / "modules"


def generate_interface(name: str, options: dict) -> str:
    if not options:
        return f"export type {name} = Record<string, unknown>;"

    lines = [f"export interface {name} {{"]
    for opt_name, opt in sorted(options.items()):
        required = opt.get("required", False)
        suffix = "" if required else "?"
        t = ts_type(opt)
        desc = (opt.get("description") or [""])[0] if isinstance(opt.get("description"), list) else (opt.get("description") or "")
        if desc:
            desc = desc.replace("*/", "* /")
            lines.append(f"  /** {desc} */")
        key = f'"{opt_name}"' if not opt_name.isidentifier() else opt_name
        lines.append(f"  {key}{suffix}: {t};")
    lines.append("}")
    return "\n".join(lines)


def generate_module(fqcn: str, doc_data: dict) -> str:
    doc = doc_data.get(fqcn, doc_data.get(list(doc_data.keys())[0], {}))
    doc_info = doc.get("doc", {})
    return_info = doc.get("return", {}) or {}

    short_name = fqcn.rsplit(".", 1)[-1]
    pascal = to_pascal(short_name)
    fn_name = safe_identifier(short_name)

    options = doc_info.get("options", {}) or {}
    args_interface = generate_interface(f"{pascal}Args", options)
    return_interface = generate_interface(f"{pascal}Return", return_info)

    return f'''// Auto-generated from: {fqcn}
// DO NOT EDIT — regenerate with codegen

import {{ defineModule }} from "@sensible-ts/core";

{args_interface}

{return_interface}

export const {fn_name} = defineModule<{pascal}Args, {pascal}Return>("{fqcn}");
'''


def main():
    if len(sys.argv) < 2:
        print(f"Usage: {sys.argv[0]} <namespace> [module...]", file=sys.stderr)
        sys.exit(1)

    namespace = sys.argv[1]
    explicit_modules = sys.argv[2:] if len(sys.argv) > 2 else None
    output_dir = namespace_to_dir(namespace)

    output_dir.mkdir(parents=True, exist_ok=True)

    if explicit_modules:
        modules = [f"{namespace}.{m}" if not m.startswith(namespace) else m for m in explicit_modules]
    else:
        modules = list_collection_modules(namespace)

    if not modules:
        print(f"No modules found for namespace: {namespace}", file=sys.stderr)
        sys.exit(1)

    generated = []
    for fqcn in modules:
        short_name = fqcn.rsplit(".", 1)[-1]
        fn_name = safe_identifier(short_name)
        print(f"  Generating {fqcn}...", file=sys.stderr)
        try:
            doc_data = run_ansible_doc(fqcn)
            source = generate_module(fqcn, doc_data)
            out_file = output_dir / f"{short_name}.ts"
            out_file.write_text(source)
            generated.append((short_name, fn_name))
        except (subprocess.CalledProcessError, KeyError, json.JSONDecodeError) as e:
            print(f"  WARNING: skipping {fqcn}: {e}", file=sys.stderr)

    all_modules = sorted(
        (p.stem, safe_identifier(p.stem))
        for p in output_dir.glob("*.ts")
        if p.stem != "index"
    )

    lines = [
        "// Auto-generated barrel — DO NOT EDIT",
        "",
    ]
    for file_name, fn_name in all_modules:
        lines.append(f'export * from "./{file_name}.ts";')
    lines.append("")
    (output_dir / "index.ts").write_text("\n".join(lines))

    print(f"Generated {len(generated)} modules in {output_dir}", file=sys.stderr)


if __name__ == "__main__":
    main()
